streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java

```diff
@@ -349,6 +359,57 @@ public String queryableStoreName() {
         return toStream().selectKey(mapper);
     }
 
+    @Override
+    public KTable<K, V> suppress(final Suppressed<K, V> suppressed) {
+        final SuppressedImpl<K, V> suppressImpl = (SuppressedImpl<K, V>) suppressed;
+        final String name = builder.newProcessorName(SUPPRESS_NAME);
+
+        final ProcessorSupplier<K, Change<V>> suppressionSupplier =
+            () -> new KTableSuppressProcessor<>(buildSuppress(suppressImpl));
+
+        final ProcessorParameters<K, Change<V>> processorParameters = new ProcessorParameters<>(
+            suppressionSupplier,
+            name
+        );
+
+        final ProcessorGraphNode<K, Change<V>> node = new ProcessorGraphNode<>(name, processorParameters, false);
+
+        builder.addGraphNode(this.streamsGraphNode, node);
+
+        return new KTableImpl<K, S, V>(
+            builder,
+            name,
+            suppressionSupplier,
+            keySerde,
+            valSerde,
+            Collections.singleton(this.name),
+            null,
+            false,
+            node
+        );
+    }
+
+    @SuppressWarnings("unchecked") // see the comments on the casts
+    private SuppressedImpl<K, V> buildSuppress(final SuppressedImpl<K, V> suppress) {
+        if (suppress.isFinalResultsSuppression()) {
+            // This throws an exception if the computation is not windowed
+            final long grace = findAndVerifyWindowGrace(streamsGraphNode);
+
+            // so we know we can cast the key to Windowed
+            final SuppressedImpl<? extends Windowed, V> windowedSuppress = (SuppressedImpl<? extends Windowed, V>) suppress;
+
+            final SuppressedImpl<? extends Windowed, V> finalResultsSuppression = SuppressedImpl.buildFinalResultsSuppression(
```

Comment:
No, the final results suppression is defined only for Windowed tables, by design.

As the comments explained, we happen to know that K is a Windowed, but the generic bound is still just `K extends Object`, so we have to "forget" our knowledge that K is a Windowed in order to return the result properly typed.

This is what I was trying to explain in the comments, and your question illustrates why I think the comments are necessary. But I guess I need to try harder to make the comments explain this situation fully.

Reply:
I understood the comments -- was just not sure if we can improve the code. If we cannot change the generic type, it's fine.

Reply:
Ah, sorry, I misunderstood the root of your question.

Since the type system doesn't have evidence that `K` extends `Windowed`, we have to do a cast to assign to `<K,V>`. I separated it into the next line just to avoid doing too much in one line of code.